src/Agent.Listener/JobDispatcher.cs

```diff
@@ -1,44 +1,231 @@
+        Task WaitAsync(CancellationToken token);
+        Task Shutdown();
     }
 
     public sealed class JobDispatcher : AgentService, IJobDispatcher
     {
+        private class Dispatch : IDisposable
+        {
+            public Dispatch(Guid jobId, long requestId)
+            {
+                JobId = jobId;
+                RequestId = requestId;
+                WorkerCancellationTokenSource = new CancellationTokenSource();
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+
+            private void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    if (WorkerCancellationTokenSource != null)
+                    {
+                        WorkerCancellationTokenSource.Cancel();
+                        WorkerCancellationTokenSource = null;
+                    }
+                }
+            }
+
+            public long RequestId { get; }
+            public Guid JobId { get; }
+            public Task WorkerDispatch { get; set; }
+            public CancellationTokenSource WorkerCancellationTokenSource { get; private set; }
+        }
+
         private static readonly string _workerProcessName = $"Agent.Worker{IOUtil.ExeExtension}";
 
+        private int _poolId;
+
+        private readonly Queue<Dispatch> _workerDispatchQueue = new Queue<Dispatch>();
+        private readonly ConcurrentDictionary<Guid, CancellationTokenSource> _workerCancellationTokenSources = new ConcurrentDictionary<Guid, CancellationTokenSource>();
+
         //allow up to 30sec for any data to be transmitted over the process channel
         private readonly TimeSpan ChannelTimeout = TimeSpan.FromSeconds(30);
 
-        public async Task<int> RunAsync(JobRequestMessage message, CancellationToken jobRequestCancellationToken)
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+
+            // get pool id from config
+            var configurationStore = hostContext.GetService<IConfigurationStore>();
+            AgentSettings agentSetting = configurationStore.GetSettings();
+            _poolId = agentSetting.PoolId;
+        }
+
+        public void Run(JobRequestMessage jobRequestMessage)
+        {
+            Trace.Info("Job request {0} received.", jobRequestMessage.JobId);
+
+            Dispatch currentDispatch = null;
+            if (_workerDispatchQueue.Count > 0)
+            {
+                currentDispatch = _workerDispatchQueue.Dequeue();
+            }
+
+            Dispatch newDisptch = new Dispatch(jobRequestMessage.JobId, jobRequestMessage.RequestId);
+
+            newDisptch.WorkerDispatch = RunAsync(jobRequestMessage, currentDispatch, newDisptch.WorkerCancellationTokenSource.Token);
+            _workerCancellationTokenSources.TryAdd(newDisptch.JobId, newDisptch.WorkerCancellationTokenSource);
```

Comment:
Why not use a ConcurrentDictionary<Guid, Dispatch> instead? There is no reason to separate the cancellation token source from the thing which actually owns it.